openssl-dynamic/src/main/c/ssl.c

```diff
@@ -334,6 +343,47 @@ static long bio_java_bytebuffer_ctrl(BIO* bio, int cmd, long num, void* ptr) {
     }
 }
 
+// This code is based on libcurl:
+// https://github.com/curl/curl/blob/curl-7_61_0/lib/vtls/openssl.c#L521
+#ifndef OPENSSL_NO_ENGINE
+/*
+ * Supply default password to the engine user interface conversation.
+ * The password is passed by OpenSSL engine from ENGINE_load_private_key()
+ * last argument to the ui and can be obtained by UI_get0_user_data(ui) here.
+ */
+static int ssl_ui_reader(UI *ui, UI_STRING *uis)
+{
+    const char *password;
+    switch (UI_get_string_type(uis)) {
+    case UIT_PROMPT:
+    case UIT_VERIFY:
+        password = (const char *) UI_get0_user_data(ui);
+        if (password != NULL && (UI_get_input_flags(uis) & UI_INPUT_FLAG_DEFAULT_PWD) != 0) {
+            UI_set_result(ui, uis, password);
+            return 1;
+        }
```

Comment:
Add a comment that fallthrough is expected (make it clear there is a missing break on purpose)? Ditto in next function.